server/src/test/java/org/jetbrains/bsp/bazel/server/sync/languages/rust/RustWorkspaceResolverTest.kt

```diff
@@ -0,0 +1,307 @@
+
+    // when
+    val packages = resolver.rustPackages(modules)
+    val (dependencies, rawDependencies) = resolver.rustDependencies(packages, modules)
+
+    // then
+    packages.size shouldBe 1
+
+    val pkg = packages[0]
+    pkg.id shouldBe rustModule.crateRoot
+    pkg.version shouldBe rustModule.version
+    pkg.edition shouldBe rustModule.edition
+    pkg.features.map { it.name } shouldContainExactlyInAnyOrder features
+    pkg.enabledFeatures shouldContainExactlyInAnyOrder features
+
+    pkg.targets shouldNotBe null
+    pkg.targets shouldBe pkg.allTargets
+    pkg.targets.size shouldBe 1
+
+    val target = pkg.targets[0]
+    target.name shouldBe rustModule.crateId
+    target.crateRootUrl shouldBe rustModule.crateRoot
+    target.kind shouldBe rustModule.kind
+    target.edition shouldBe rustModule.edition
+    target.requiredFeatures shouldContainExactlyInAnyOrder features
+
+    dependencies shouldNotBe null
+    dependencies shouldBe emptyList()
+
+    rawDependencies shouldNotBe null
+    rawDependencies shouldBe emptyList()
+  }
+
+  @Test
+  fun `should return empty dependency list for empty package list`() {
+    // given
+    val modules = emptyList<Module>()
+    val packages = emptyList<RustPackage>()
+
+    // when
+    val (dependencies, rawDependencies) = resolver.rustDependencies(packages, modules)
+
+    // then
+    dependencies shouldNotBe null
+    dependencies shouldBe emptyList()
+
+    rawDependencies shouldNotBe null
+    rawDependencies shouldBe emptyList()
+  }
+
+  @Test
+  fun `should return proper dependency list for a target with one dependency`() {
+    // given
+    //   B
+    //   |
+    //   A
+    val rustModuleA = createRustModule(
+        crateId = "TargetA",
+        crateRoot = "/path/to/target/a",
+        dependencies = emptyList()
+    )
+    val moduleA = createModule(
+        label = "/path/to/target/a:TargetA",
+        directDependencies = emptyList(),
+        rustModule = rustModuleA
+    )
+
+    val rustModuleB = createRustModule(
+        crateId = "TargetB",
+        crateRoot = "/path/to/target/b",
+        dependencies = listOf(
+            RustDependency(
+                crateId = rustModuleA.crateId,
+                rename = rustModuleA.crateId
+            )
+        )
+    )
+    val moduleB = createModule(
+        label = "/path/to/target/b:TargetB",
+        directDependencies = listOf(moduleA.label),
+        rustModule = rustModuleB
+    )
+
+    val modules = listOf(moduleA, moduleB)
+    val packages = resolver.rustPackages(modules)
+
+    // when
+    val (dependencies, rawDependencies) = resolver.rustDependencies(packages, modules)
+
+    // then
+    dependencies.size shouldBe 1
+
+    val dependency = dependencies[0]
+    dependency.source shouldBe rustModuleB.crateRoot
+    dependency.target shouldBe rustModuleA.crateRoot
+    dependency.name shouldBe rustModuleA.crateId
+
+    rawDependencies.size shouldBe 1
+    // TODO are rawDependencies needed?
```

Comment:
Currently, raw dependencies represent the dependencies considering all targets within a single package.

If two targets are conflicting, then one of them is being removed. The dependencies represent the actual state of dependencies with the targets resolved. On the other hand raw dependencies represent all dependencies of all targets. It is not how Cargo understands it, but this is how it is implemented.